Added server constructor with custom options

zebox opened the pull request:
As we discussed early i added server options (include custom middlewares #8) and `NewServer` constructor. Also I fixed test for work with constructor.

But I am confused for a few things:

* `author` field should be always static
* `version`field it should be configurable? 
* `ClientLimit` in Limits struct used only in test, it's normal? I can't find usage anywhere else.
* assignation an `API` option, in current implementationn look like ugly. Maybe is add `API` in constructor and change it to private?

I didn't change `Readme`, because think do it after all you remarks.
I hope what i understood all as you mentioned in previous my PR.

umputun commented:
> author field should be always static

nothing wrong with setting author as in practical use cases this would be some custom service using jrpc for the communication

> version field it should be configurable?

 yes, we don't know what the version is, because this is a library and not a stand-alone application

> ClientLimit in Limits struct used only in test, it's normal?

not really. it used in tollbooth middleware `tollbooth_chi.LimitHandler(tollbooth.NewLimiter(s.Limits.ClientLimit, nil))`

> assignation an API option, in current implementationn look like ugly. Maybe is add API in constructor and change it to private?

yeah, everything should  be private now and all mandatory params should be passed as constructor's params

umputun commented:
btw, hiding signature optionally can be a good idea here, see https://github.com/umputun/reproxy/blob/master/app/proxy/handlers.go#L105

zebox commented:
> not really. it used in tollbooth middleware tollbooth_chi.LimitHandler(tollbooth.NewLimiter(s.Limits.ClientLimit, nil))

It's mean that I did wrong. `WithLimit` option should use with tollbooth middleware, but I implemented as server timeout. For server timeout need add additional option `Timeouts`. Is it right?

zebox commented:
I made change with:
* `WithLimit` options
* added `Timeouts` and `WithoutSignature` options
* make `API` private and added it to constructor params
* changed test for new options

After review the code by you I add change to `readme`.

zebox commented:
I done all changes:
* added constructor with options values
* changed `Readme`
* changed test and example

umputun commented on options.go:
```diff
@@ -0,0 +1,85 @@
+}
+
+// AppName sets custom application name for server signature, optional
+func AppName(name string) Option {
```
Probably it makes sense to add author the same way

zebox replied:
What's default value should be define for `author` field, empty or '_umputun_' ?

umputun replied:
empty should be fine

umputun commented on options.go:
```diff
@@ -0,0 +1,85 @@
+
+// Timeouts sets server timeout values such as ReadHeader, Write and Idle timeout, optional.
+// If this option not defined server use default timeout values
+func Timeouts(limits Limits) Option {
```
I think, it will be better to rename Limits here to Timeouts to avoid confusions, as we have other limits already

zebox replied:
Should is stay type as `Limits` or define new type (e.g. `Timeouts`) with fields for timeouts only?

umputun commented on options.go:
```diff
@@ -0,0 +1,85 @@
+}
+
+// WithTimeout sets middleware with custom timeout value, optional
+func WithTimeout(timeout time.Duration) Option {
```
this is confusing name as we already have Timeouts option. Maybe this on should moved to limits structure and set by the same Timeouts

zebox replied:
Yes. I thought about this and it doesn't like me too. I will move it to `Timeouts`

umputun commented on go.mod:
```diff
@@ -1,15 +1,14 @@
 module github.com/go-pkgz/jrpc
 
-go 1.14
+go 1.15
```
I think we can bump it to 1.16

zebox replied:
Yes. I will fixed it.

umputun reviewed (changes requested):
thx, a few minor things. And we also need to cover all the options use with tests

zebox commented:
I added changes with you remarks and refactor code:

1. Grouped signature data to struct `signaturePayload`. The field `signature` in `Server` struct use as pointer for checks all fields at one time. Signature by default is disabled, because default value for `appName`, `author` and `version` is empty.
2. Removed `Timeouts` option and implement timeouts defined in option `WithTimeouts`. I splited `limits` and `timeouts` to different struct.
3. Added tests for all options (may be `WithLogger` test looks not good)

umputun reviewed (commented):
LGTM. I see a few small things i would like to adjust but will do it after the merge. Pls squash all the commints and I'll merge it in

zebox commented:
> LGTM. I see a few small things i would like to adjust but will do it after the merge. Pls squash all the commints and I'll merge it in

Done!
